node/cn/peer_set_test.go

```diff
@@ -0,0 +1,533 @@
+	assert.EqualValues(t, expectedPeers, peers)
+}
+
+func TestPeerSet_CNPeers(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cnPeer := NewMockPeer(mockCtrl)
+	cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(1)
+	setMockPeers([]*MockPeer{cnPeer})
+
+	assert.EqualValues(t, map[common.Address]Peer{}, peerSet.CNPeers())
+	assert.NoError(t, peerSet.Register(cnPeer))
+	assert.EqualValues(t, map[common.Address]Peer{addrs[0]: cnPeer}, peerSet.CNPeers())
+}
+
+func TestPeerSet_PNPeers(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	pnPeer := NewMockPeer(mockCtrl)
+	pnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.PROXYNODE)).Times(1)
+	setMockPeers([]*MockPeer{pnPeer})
+
+	assert.EqualValues(t, map[common.Address]Peer{}, peerSet.PNPeers())
+	assert.NoError(t, peerSet.Register(pnPeer))
+	assert.EqualValues(t, map[common.Address]Peer{addrs[0]: pnPeer}, peerSet.PNPeers())
+}
+
+func TestPeerSet_ENPeers(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	enPeer := NewMockPeer(mockCtrl)
+	enPeer.EXPECT().ConnType().Return(p2p.ConnType(node.ENDPOINTNODE)).Times(1)
+	setMockPeers([]*MockPeer{enPeer})
+
+	assert.EqualValues(t, map[common.Address]Peer{}, peerSet.ENPeers())
+	assert.NoError(t, peerSet.Register(enPeer))
+	assert.EqualValues(t, map[common.Address]Peer{addrs[0]: enPeer}, peerSet.ENPeers())
+}
+
+func TestPeerSet_Peer_And_Len(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cnPeer := NewMockPeer(mockCtrl)
+	pnPeer := NewMockPeer(mockCtrl)
+	enPeer := NewMockPeer(mockCtrl)
+
+	setMockPeersConnType(cnPeer, pnPeer, enPeer)
+	setMockPeers([]*MockPeer{cnPeer, pnPeer, enPeer})
+
+	assert.Equal(t, 0, peerSet.Len())
+	assert.NoError(t, peerSet.Register(cnPeer))
+	assert.Equal(t, 1, peerSet.Len())
+	assert.NoError(t, peerSet.Register(pnPeer))
+	assert.Equal(t, 2, peerSet.Len())
+	assert.NoError(t, peerSet.Register(enPeer))
+	assert.Equal(t, 3, peerSet.Len())
+
+	assert.Equal(t, cnPeer, peerSet.Peer(nodeids[0].String()))
+	assert.Equal(t, pnPeer, peerSet.Peer(nodeids[1].String()))
+	assert.Equal(t, enPeer, peerSet.Peer(nodeids[2].String()))
+}
+
+func TestPeerSet_PeersWithoutBlock(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	block := newBlock(blockNum1)
+
+	cnPeer := NewMockPeer(mockCtrl)
+	pnPeer := NewMockPeer(mockCtrl)
+	enPeer := NewMockPeer(mockCtrl)
+
+	setMockPeersConnType(cnPeer, pnPeer, enPeer)
+	setMockPeers([]*MockPeer{cnPeer, pnPeer, enPeer})
+
+	cnPeer.EXPECT().KnowsBlock(block.Hash()).Return(false).AnyTimes()
+	pnPeer.EXPECT().KnowsBlock(block.Hash()).Return(true).AnyTimes()
+	enPeer.EXPECT().KnowsBlock(block.Hash()).Return(false).AnyTimes()
```

Comment:
Why don't you test returning true and false for all cn/pn/en peers?

Reply:
If you are asking the reason why `AnyTimes()` is used instead of `Times()`, I changed it since what we are testing here is the result of `PeersWithoutBlock`, not the number of times `Times()` is called. Regardless of the call numbers, `PeersWithoutBlock` should return the same result. However, the number of times `KnowsBlock(block.Hash())` is not our concern. It can be called any times if the internal implementation of `PeersWithoutBlock` is changed.

Reply:
I meant that we can test the cases below. (These cases are not considered in present code.)
- `CNWithoutBlock` returns empty peer
- `PNWithoutBlock` returns a peer
- `ENWithoutBlock` returns empty peer